server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java

```diff
@@ -26,4 +26,6 @@
 public interface UserProfileProvider extends Provider {
 
     UserProfileValidationResult validate(UserProfileContext updateContext, UserProfile updatedProfile);
+
+    boolean isReadOnlyAttribute(String key);
```

Comment:
I am not 100% sure about adding "boolean isReadOnlyAttribute(String key);" as the first-level citizen method to the UserProfileProvider interface. It seems that the fact if attribute is readOnly may depend on who is updating it, and maybe sometimes also when is updating it (EG. username can be created just during registration, but not updated during account console editation etc).

This is not a blocker for this bugfix as UserProfileProvider is private SPI and we can change it later. But just something to consider for the future UserProfile work...

Reply:
That is why we have that method wrapped by org.keycloak.userprofile.UserProfileAttributes#isReadOnlyAttribute. Which is what should provide more contextual information for attributes.

This is one of the things we need to improve in the SPI, so that we can bring the context closer to attributes. I did not want to change much to not lose too much focus on the problem being fixed. The read-only checks should come dynamically and not hard-coded as it stands today, from the profile config.